### PR TITLE
mrc-4758 Fix download file when authenticated

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch fixes the download file links when auth is enabled and user is logged in - previously this was failing because we were rendering plain `<a>` links to the downloads, which of course to do not pass the required `Authorization` header. 

I've added a `download` lib method which passes the header if required, and spoofs a link click to the actual download. 
I've replaced the previous links with a `DownloadButton` component - and made this look more like a button than a link.
This also shows an error message if there is an error downloading the file. 

NB this PR is merging to the theme update epic branch in order to pick up the new `lib` structure and msw support.

### CODE REVIEW

File: .github/workflows/backend-test-and-build.yml
Comment (absternator) [inline .github/workflows/backend-test-and-build.yml:7]: Might need to do another pull from 4716

File: app/src/app/components/contents/download/DownloadButton.tsx
Comment (absternator) [inline app/src/app/components/contents/download/DownloadButton.tsx:16]: pretty sure you dont have check authenicated here.... as you have to be authenticated to see this page

File: app/src/lib/download.ts
Comment (absternator) [inline app/src/lib/download.ts:4]: can use existing fetcher to achieve all this data fetching pretty sure....

File: app/src/lib/download.ts
Comment (absternator) [inline app/src/lib/download.ts:16]: try catch around this

File: app/src/lib/download.ts
Comment (absternator) [inline app/src/lib/download.ts:17]: sorry just my ignorance but is this the only way to do this? seems really hacky for some reason

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (absternator) [inline app/src/tests/components/contents/download/DownloadButton.test.tsx:22]: as above dont think we neeed to test if user is authenticated or not

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (absternator) [inline app/src/tests/components/contents/download/DownloadButton.test.tsx:1]: maybe best to move under imports

File: app/src/tests/lib/auth/getAuthHeader.test.ts
Comment (absternator) [inline app/src/tests/lib/auth/getAuthHeader.test.ts:1]: below imports

File: app/src/tests/lib/auth/getAuthHeader.test.ts
Comment (absternator) [inline app/src/tests/lib/auth/getAuthHeader.test.ts:3]:
```suggestion
    getBearerToken: jest.fn(() => "1234")
```
Can avoid creating another var

File: app/src/tests/lib/download.test.ts
Comment (absternator) [inline app/src/tests/lib/download.test.ts:16]: shouldnt need to mock fetch... can use msw handler to intercept request.. msw purpose is so we can avoid this :smile:

**Review by absternator — COMMENTED** [review]
LGTM the code is sweet as just some minor comments... Havent tested yet il wait till parent branch is merged in as lots of changes

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:1]: I think it has to go before the import so that the real download doesn't get pulled in by the imported method - will confirm this though,

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:22]: This is because we might be running an instance with auth switched off.

File: app/src/tests/lib/auth/getAuthHeader.test.ts
Comment (EmmaLRussell) [reply app/src/tests/lib/auth/getAuthHeader.test.ts:3]: Ah yeah, can do here actually since we don't need to explicitly expect the mock

File: app/src/tests/lib/download.test.ts
Comment (EmmaLRussell) [reply app/src/tests/lib/download.test.ts:16]: I want to spy on  it though... :mag:

Comment (EmmaLRussell) [reply app/src/tests/lib/download.test.ts:16]: Unless msw has some spy magic I don't know about?

File: app/src/lib/download.ts
Comment (EmmaLRussell) [reply app/src/lib/download.ts:17]: It seems to be sadly... Yeah, that was my feeling too, but this seems to be the standard approach.

File: app/src/lib/download.ts
Comment (EmmaLRussell) [reply app/src/lib/download.ts:4]: Yeah, I wondered about that, but would need to modify the existing fetcher, which currently just returns JSON. I figured it wouldn't actually share very much code other than appending the auth header (since this has to deal with a blob on success and json on failure). I could put in the capacity to the current fetcher to allow blob return but I think that would overcompilicate that probably..

File: .github/workflows/backend-test-and-build.yml
Comment (EmmaLRussell) [reply .github/workflows/backend-test-and-build.yml:7]: Cheers, yep will repoint the PR once 4716 merged to main,

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:22]: Actually, we could just let the getAuthHeader method return nothing if there's no current user, that might be simpler as the clients of the function wouldn't need to know, but might be confusing since the header returned might be empty,

File: app/src/lib/download.ts
Comment (absternator) [reply app/src/lib/download.ts:4]: ohh didnt realize it does a blob on the res directly.... yup good

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (absternator) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:1]: it wont.. when you use jest.mock it auto gets hoiste to top of file :smile:

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (absternator) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:22]: ohh yeah that auth instance off scenario

File: app/src/tests/lib/download.test.ts
Comment (absternator) [reply app/src/tests/lib/download.test.ts:16]: what exactly are you spying on? is the direct spyon fetch needed for the test? looks like it line 48 is the assertion... unsure if that is needed since you are already checking lots of assertions afterwards?

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:1]: Yeah, that's what I thought it should do, but I'm sure it's been a problem in the past and only setting up mocks pre-imports fixed it. Will dig into that though, you're right, shouldn't need to do that.

File: app/src/tests/lib/download.test.ts
Comment (EmmaLRussell) [reply app/src/tests/lib/download.test.ts:16]: I think I wanted to use the spy to check the header had been sent correctly, and then I needed to restore the real fetch for subsequent fetches to invoke msw. But actually I could probably configure msw to only kick in when the correct header is provided, to return the mock blob, and that would confirm that the header had been sent.. I'll give that a try.

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx:22]: It can be downloaded if there's no header and auth is switched off on the server, but you'll get a 401 if auth is switched on and no valid header is provided. But in terms of front end testing, we can just check that we provide the token if we have one.

File: app/src/tests/components/contents/download/DownloadButton.test.tsx
Comment (EmmaLRussell) [reply app/src/tests/components/contents/download/DownloadButton.test.tsx]: It was fine! Must have been having other issues when I did this before! 😆

**Review by absternator — APPROVED** [review]

**Pull request merged**
